tests.unit/test_perf_report_generator.py

```diff
@@ -0,0 +1,5072 @@
+  gFmpDevicePkgTokenSpaceGuid.PcdFmpDeviceImageTypeIdGuid|{0}|VOID*|0x40000010
+
+[PcdsFixedAtBuild, PcdsPatchableInModule, PcdsDynamic, PcdsDynamicEx]
+  ## One or more PKCS7 certificates used to verify a firmware device capsule
+  #  update image.  Encoded using the Variable-Length Opaque Data format of RFC
+  #  4506 External Data Representation Standard (XDR).  The default value is
+  #  empty with 0 certificates.
+  # @Prompt One or more XDR encoded PKCS7 certificates used to verify firmware device capsule update images.
+  gFmpDevicePkgTokenSpaceGuid.PcdFmpDevicePkcs7CertBufferXdr|{0x0}|VOID*|0x4000000E
+
+  ## An event GUID that locks the firmware device when the event is signaled.
+  #  If this PCD is not a valid GUID value, then the firmware device is locked
+  #  when gEfiEndOfDxeEventGroupGuid (End of DXE Phase) is signaled.  The
+  #  default value is empty, so by default the firmware device is locked at the
+  #  end of the DXE phase.
+  # @Prompt Firmware Device Lock Event GUID.
+  gFmpDevicePkgTokenSpaceGuid.PcdFmpDeviceLockEventGuid|{0}|VOID*|0x4000000F
+
+[UserExtensions.TianoCore."ExtraFiles"]
+  FmpDevicePkgExtra.uni
+"""
+
+
+# Simple FDF file content (OVMF/QEMU snippets)
+FdfTestModule = namedtuple("FdfTestModule", ["guid", "name"])
+
+FDF_TEST_MODULE = FdfTestModule("12345678-1234-1234-1234-123456789ABC", "TestModule")
+FDF_TEST_MODULE2 = FdfTestModule("ABCDEF12-3456-7890-1234-56789ABCDEF0", "TestModule2")
+
+QEMU_FDF_CONTENT = f"""## @file
+#  Open Virtual Machine Firmware for Q35 VM using Project Mu: FDF
+#
+#  Copyright(c) Microsoft Corporation
+#  Copyright (c) 2006 - 2019, Intel Corporation. All rights reserved.<BR>
+#  (C) Copyright 2016 Hewlett Packard Enterprise Development LP<BR>
+#
+#  SPDX-License-Identifier: BSD-2-Clause-Patent
+#
+##
+
+################################################################################
+
+[Defines]
+DEFINE BLOCK_SIZE        = 0x1000
+
+DEFINE VARS_SIZE         = 0x84000
+DEFINE VARS_BLOCKS       = 0x84
+DEFINE VARS_LIVE_SIZE    = 0x40000
+DEFINE VARS_SPARE_SIZE   = 0x42000
+
+DEFINE FW_BASE_ADDRESS   = 0xFFC00000
+DEFINE FW_SIZE           = 0x00400000
+DEFINE FW_BLOCKS         = 0x400
+DEFINE CODE_BASE_ADDRESS = 0xFFC84000
+DEFINE CODE_SIZE         = 0x0037C000
+DEFINE CODE_BLOCKS       = 0x37C
+DEFINE FVMAIN_SIZE       = 0x00348000
+DEFINE SECFV_OFFSET      = 0x003CC00
+DEFINE SECFV_SIZE        = 0x34000
+
+SET gUefiQemuQ35PkgTokenSpaceGuid.PcdOvmfFdBaseAddress     = $(FW_BASE_ADDRESS)
+SET gUefiQemuQ35PkgTokenSpaceGuid.PcdOvmfFirmwareFdSize    = $(FW_SIZE)
+
+[FD.QEMUQ35_VARS]
+BaseAddress   = $(FW_BASE_ADDRESS)
+Size          = $(VARS_SIZE)
+ErasePolarity = 1
+BlockSize     = $(BLOCK_SIZE)
+NumBlocks     = $(VARS_BLOCKS)
+
+0x00000000|0x00040000
+#NV_VARIABLE_STORE
+DATA = {
+    ## This is the EFI_FIRMWARE_VOLUME_HEADER
+    # ZeroVector []
+    (
+        0x00,
+        0x00,
+        0x00,
+        0x00,
+        0x00,
+        0x00,
+        0x00,
+        0x00,
+    )
+}
```

Comment:
This `}` needs to be escaped so python does not think you are trying to do f-string replacement here.